### PR TITLE
Adds indeterminate state to checkbox

### PR DESCRIPTION
**Pull request by n1colemejia**
## Summary:
Adds an indeterminate state to checkbox component by setting `checked` to `null`/`undefined`. When `checked` is `null`/`undefined`, checkbox will show an indeterminate icon :heavy_minus_sign:. 

Issue: https://khanacademy.atlassian.net/browse/WB-1460

## Test plan:
- `yarn jest` `yarn typecheck`
- In the [Indeterminate story](https://5e1bf4b385e3fb0020b7073c-obfmyxbzyo.chromatic.com/?path=/story/form-checkbox--indeterminate) for Checkbox, inspect the checkboxes in default, disabled and error state to see the correct indeterminate icon appears and that the `aria-checked` attribute is `mixed`.

  Screenshot 2023-03-28 at 12 39 06 PM

### CODE REVIEW

**Comment by jandrade** [comment]
@n1colemejia I think that indeterminate states are not directly connected to the `checked` state. You can see more about that here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes

> In addition to the checked and unchecked states, there is a third state a checkbox can be in: indeterminate. This is a state in which it's impossible to say whether the item is toggled on or off.

I'm not sure if this was discussed before, so in case of that, I'd like to open the conversation and get your thoughts about this.

File: packages/wonder-blocks-form/src/components/checkbox.tsx
Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/checkbox.tsx:11]: When `checked: boolean | null | undefined` in `checkbox.tsx`, I get type errors that say `checked` is incompatible: 

Screenshot 2023-03-31 at 10 27 18 AM
Screenshot 2023-03-31 at 10 28 30 AM

I changed the `checked` type to match in `choice.tsx`, `choice-internal.tsx`, and `types.ts` and the error didn't go away.

Also recieved type error in `checkbox-core.tsx` for `StyledInput` `checked` prop that led to `index.d.ts` for `InputHTMLAttributes` where `checked?: boolean | "mixed" | undefined`.

Screenshot 2023-03-31 at 10 29 59 AM

In Storybook:

The indeterminate icon appears when `checked={null}` or `checked="mixed"`, but doesn't appear when `checked={undefined}`.

The `aria-checked` doesn't appear in the DOM when the `checked` value is `null` or `undefined`, only for `checked="mixed"`

I could totally by missing something or going about this wrong! This is just the scavenger hunt that led me to settle on `boolean | "mixed"`, but I agree that `null`/`undefined` feels like a better approach.

Comment (jandrade) [reply packages/wonder-blocks-form/src/components/checkbox.tsx:11]: ahh ok, thanks for all the context, it's super helpful :).

So the issue that you experienced was because, I suspect there's one missing step to be able to use `null/undefined` in checkboxes.

The `props.checked` value can still have `true | false | null | undefined`, but we'd need to map that value to `aria-checked` so that ARIA attribute can have the right value and avoid that TS error.

I could imagine something like this would work (not fully sure)...

Just adding a snippet for illustration purposes:

```
type Checked = boolean | null | undefined;
type AriaChecked = "true" | "false" | "mixed";

function mapCheckedToAriaChecked(value: Checked): AriaChecked {
  switch(value) {
    case true:
      return "true";
    case false:
      return "false";
    default:
      return "mixed";
  }
}

// then in the component
const ariaChecked = mapCheckedToAriaChecked(props.checked);
...
return (
  <input type="checkbox" aria-checked={ariaChecked} ... />
);
```

What do you think? happy to chat more about it.

Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/checkbox.tsx:11]: Oh awesome, thank you Juan! I will try this and check back to see about pairing 😊

Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/checkbox.tsx:11]: So that worked! The only issue is that `undefined` sets `aria-checked` to `false` instead of `"mixed"` even though it "looks" right.

**Review by megaroeny — APPROVED** [review]
The new story looks great Nicole!! 🎉

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:11]: **suggestion:** this type can be exported so it can be reused in `checkbox.tsx`. It help us to centralize these types in case there's a future change:

```suggestion
export type Checked = boolean | null | undefined;
```

Comment (jandrade) [reply packages/wonder-blocks-form/src/components/checkbox-core.tsx:11]: Also, it would be really helpful adding a comment about what `null | undefined` means in this context (or more specifically that this intends to be used when the checkbox is in a indeterminate state.

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:15]: **suggestion:** It would be also helpful adding comments here about the why we define this intermediate step... or describing the purpose of this function (e.g. expressing that we convert the prop value to a value that a Screen Reader can understand via the `aria-checked` attribute).

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:86]: **question:** Do you think we could fix this by merging your changes from #1817 (after that PR is landed)? so we could solve this right away :).

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:167]: **praise:** This is a great use of reusability here!

File: packages/wonder-blocks-form/src/components/checkbox.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox.tsx:11]: **suggestion:** we could use the `Checked` type exported from `checkbox-core.tsx` here.

File: packages/wonder-blocks-form/src/components/radio-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/radio-core.tsx:32]: **question:** does this happen b/c the `checked` type differs from the one in `ChoiceInternal/Checkbox`?

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:92]: **suggestion:** you can get rid of this `role` as it is implicitly defined by the `input[checkbox]` element.

```suggestion         
```

File: packages/wonder-blocks-form/src/__tests__/__snapshots__/custom-snapshot.test.tsx.snap
Comment (jandrade) [inline packages/wonder-blocks-form/src/__tests__/__snapshots__/custom-snapshot.test.tsx.snap:5]: **suggestion:** that makes me realize that we should add a new variant to cover the `indeterminate` case.

This test can be included here: https://github.com/Khan/wonder-blocks/blob/c2ec490259d77ec44dcd6b868d050314c93d72e9/packages/wonder-blocks-form/src/__tests__/custom-snapshot.test.tsx#L9

File: packages/wonder-blocks-form/src/util/types.ts
Comment (jandrade) [inline packages/wonder-blocks-form/src/util/types.ts:12]: **thought:** Giving a second thought about from where we should export `Checked`... maybe this is the best place from where we should define the `Checked` type as it is the shared/central place for both checkboxes and radiobuttons. wdyt?

**Review by jandrade — COMMENTED** [review]
Thanks for working on this, great progress!! Things look really well, but I'd like to get your thoughts in a couple of points before giving final approval.

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/checkbox-core.tsx:86]: Yes I like that idea! I landed #1817 but I can't merge the changes until Version Packages are merged, right?

File: packages/wonder-blocks-form/src/components/radio-core.tsx
Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/radio-core.tsx:32]: The error goes away when I update the type for `checked` to `null | undefined` where `_generateStyles` is defined later in the file. 

Not sure if this makes sense though since `Radio` doesn't have an indeterminate state. Should the `checked` type for `Checkbox`/`Radio` be the same?

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [reply packages/wonder-blocks-form/src/components/checkbox-core.tsx:86]: good question! Yes, you can merge changes right now :)

You can see it here: https://github.com/Khan/wonder-blocks/blob/main/packages/wonder-blocks-core/src/util/add-style.tsx#L48

-----------

For more context, you could say that releasing features are a two-step process: 

1. The "code implementation" changes for a particular PR are merged into `main` right after a PR is landed.
2. The version updates for a particular package are merged into `main` right after `Version Packages` are merged.

This means that, in (1) all the code associated to your PR is directly merged into the main branch as soon as these are landed, and this code is available inside the `wonder-blocks` repo, but it's not available for consumers (like the `webapp` repo) yet. That's where (2) comes into the process, because that's where we make these changes available to consumers by releasing the package changes to npm.

File: packages/wonder-blocks-form/src/components/radio-core.tsx
Comment (jandrade) [reply packages/wonder-blocks-form/src/components/radio-core.tsx:32]: Ahh got it. I think it's should be fine to change it to `Checked` as `RadioCore` is already using the props defined in `ChoiceInternal`.

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (n1colemejia) [reply packages/wonder-blocks-form/src/components/checkbox-core.tsx:86]: Oh I see! Thank you so much for the context. That makes more sense to me now.

File: packages/wonder-blocks-form/src/components/checkbox-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/checkbox-core.tsx:12]: **suggestion:** Move this to `/util/types.ts` so you can reference this type everywhere.

File: packages/wonder-blocks-form/src/components/choice-internal.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/choice-internal.tsx:15]: **suggestion:** By moving `Checked` to `types.ts`, you can do `import type {Checked} from "./util/type"`, then:

```suggestion
    checked: Checked;
```

File: packages/wonder-blocks-form/src/components/radio-core.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/components/radio-core.tsx:117]: Same here:

```suggestion
    checked: Checked,
```

File: packages/wonder-blocks-form/src/util/types.ts
Comment (jandrade) [inline packages/wonder-blocks-form/src/util/types.ts:12]: And here:

```suggestion
    checked: Checked;
```

File: packages/wonder-blocks-form/src/__tests__/custom-snapshot.test.tsx
Comment (jandrade) [inline packages/wonder-blocks-form/src/__tests__/custom-snapshot.test.tsx:8]: **nit:** This will evaluate to `undefined`, so you can probably use `null` here to be more explicit the indeterminate case:

```suggestion
const checkedStates = [false, true, null];
```

P.S. Don't forget to update the snapshots :).

File: packages/wonder-blocks-form/src/util/types.ts
Comment (n1colemejia) [reply packages/wonder-blocks-form/src/util/types.ts:12]: I agree! That makes more sense to me as well. Especially since we're using the same type for radio

**Review by jandrade — APPROVED** [review]
Looks really great! thanks for applying the suggestions 🚀.

**Pull request merged**
